### ISSUE TITLE
Problem in loading shared object file

### ISSUE THREAD

**Issue by lee-zion** [usnistgov/REFPROP-cmake#36]
Hi.

I builded fortran source code in docker (ubuntu 18.04), and got .so file.
When I load REFPROP, it returns error like
```
'Could not load REFPROP (librefprop.so) due to: libgfortran.so.4: cannot open shared object file: No such file or directory. Make sure the library is in your system search path. In case you run 64bit Windows and you have a REFPROP license, try installing the 64bit DLL from NIST. '
```

I builded source code in my local machine(mac 10.14.6), and got .dylib file which works fine.
I searched and found one that seems related to my problem.

https://stackoverflow.com/questions/46516394/how-to-install-libgfortran-so-4-on-ubuntu-16-06/46516499

This post suggests install GCC 7.

I've tried, but the error remains.
Can you provide any suggestions for debugging?

Thanks.

**Comment by ianhbell** [comment]
The .so you built in your docker container (how did you do so?) can only be definitely used in your docker container, and *might* work outside if you can ensure binary compatibility, but that is is a big *if*.

What code did you run to get this error?

In short, build the shared library on the machine you plan to run your code on.

**Comment by lee-zion** [comment]
Here's what I'm understanding now.

This repository 'REFPROP-cmake' build shared library according to its build environment.
I just followed your READMD.md in /docker/ubuntu-18.04.
By running cmake in docker (ubuntu 18.04), built librefprop.so without compile error.
And by running cmake in OS X (10.14.6), built librefprop.dylib without compile error.

```
    Env env = info.Env();
    std::string absPath = info[0].As<String>().Utf8Value();
    std::string dllName = info[1].As<String>().Utf8Value();
    std::string err;
    bool loaded_REFPROP = load_REFPROP(err, absPath, dllName);
    if (!loaded_REFPROP)
    {
        Napi::Value retValue = Napi::String::New(env, err.c_str());
        return retValue.ToObject();
    }
   ...
```

Code executes error at loading, error message with
```
'Could not load REFPROP (librefprop.so) due to: 
libgfortran.so.4: cannot open shared object file: No such file or directory.
Make sure the library is in your system search path.
In case you run 64bit Windows and you have a REFPROP license, try installing the 64bit DLL from NIST. '
```

And I'm running this code both inside docker(ubuntu) and OS X. I thought that I built shared library with same process, and running environment, too.

Dockerfile that runs the library installs as below

```
RUN apt-get -y -m update && \
    apt-get install -y cmake g++ python3 curl
RUN curl -sL https://deb.nodesource.com/setup_10.x | bash
RUN apt-get install -y nodejs
```

Should I install something more to run shared library? (like gfortran or something?)

**Comment by ianhbell** [comment]
Please show your entire runnable (failing) codes, not just snippets.  Otherwise I have no idea how to debug remotely.

**Comment by lee-zion** [comment]
I understand runnable code means all the source code, so I made little repo, but some functions are removed and simplified.

Here is [Link](https://github.com/lee-zion/REFPROP-debug) to repo.

I've wrote responsibility of directories in README, so please let me know what you need more to help.

Summarizing for my status, produce-lib directory produces shared library file, and is just a copy of this REFPROP-cmake repo.
I've built in both my local machine(OS X) and docker, then got librefprop.dylib and librefprop.so, respectively. Both files are copied to produce-node for testing.

I wrote C++ code to produce node file, and javascript code to test node file and shared library file both. Test JS code and C++ code is in produce-node directory.

All compiled fine without error message, but JS test code returns error message as I mentioned.

```
'Could not load REFPROP (librefprop.so) due to: 
libgfortran.so.4: cannot open shared object file: No such file or directory.
Make sure the library is in your system search path.
In case you run 64bit Windows and you have a REFPROP license, try installing the 64bit DLL from NIST. '
```

And I'm wondering if that `libgfortran.so.4: cannot ...` message is from lack of package in docker, or something else that I don't recognize. 

I've checked that C++ code works(I can run test code without error in OS X and Windows), and test returns error message only in Docker(Ubuntu).

**Comment by ianhbell** [comment]
It looks like you are using librefprop.so outside the docker container where it was built?  That's a no-no.  The shared library can only be used inside the container where it was built.  For node, you should add the REFPROP shared library building to the setup of your docker container.  How does node handle loading of shared libraries?  You might need to pass the absolute path to the shared library.

**Comment by ianhbell** [comment]
Probably unrelated: you are not passing strings correctly.  REFPROP uses fixed-length strings without null-termination, and not the null-terminated strings of C/C++.  Therefore, calls like this ``const_cast<char *>(fldName.c_str())`` might work, or maybe not.  To be sure, you should copy the string into a space-padded buffer and remove the null-termination character. Something like this: https://github.com/usnistgov/REFPROP-tests/blob/master/include/REFPROPtests/baseclasses.h#L71-L75

**Comment by ianhbell** [comment]
A test that you should run before jumping into node is to get the C++ code running inside a docker container, but without all the node stuff.

**Comment by lee-zion** [comment]
1. > The shared library can only be used inside the container where it was built.

I've used Ubuntu 18.04 for both building shared library and node, but it doesn't solve this one?
Then should I just build library and node at a single docker container?

2. > For node, you should add the REFPROP shared library building to the setup of your docker container. 

This means I should build library and node at a single docker container. Right? BTW, I'm using Heroku as SaaS and they use ubuntu for serving. The only way to ensure that node and library work is to run integrated docker(like Question 1) inside other's server?

3. > How does node handle loading of shared libraries?

Loading shared libraries are done in C++ from REFPROP-Wrapper, and node executes C++ code with node-addon-api (I don't know the exact way how C++ handles shared libraries. What I think is REFPROP-Wrapper handles the loading shared library in C++, and node-addon-api module handles executing C++ in node.).

4. > You might need to pass the absolute path to the shared library.

I treated path as input parameter (from JS) in C++, and JS get project's current absolute path from

```
var nodepath = path.join(__dirname, '..', 'docker-refprop.node')
var libpath = path.join(__dirname, '..', 'refprop', 'librefprop.so')
refprop = require(nodepath)
...
refprop.getProp(nodepath, libpath, ...)
```
What you mean is this way of passing path is wrong?

5. > A test that you should run before jumping into node is to get the C++ code running inside a docker container, but without all the node stuff.

I developed this one from C++ code(node-addon-api for node and REFPROP-Wrapper for REFPROP) and compiled with node-gyp, then got JS test node.

> I've checked that C++ code works(I can run test code without error in OS X and Windows), and test returns error message only in Docker(Ubuntu).

What I've meant to do both at all is just to save time for testing C++ code, since my running environment will be JS. Is there any problem in this procedure? I compile as all C++ code does, so I thought there would be no problem (consequently there seems no issue in C++ code since I can run in Windows or OS X.).  

What you mean is the first step should be C++ code with REFPROP-Wrapper only, then add node-addon-api code. Am I right?

6. > Probably unrelated: you are not passing strings correctly.

Thank you so much for your reviewing. It just works, but your comment got me to make much stable code :)

**Comment by ianhbell** [comment]
1. No, because you cannot guarantee ABI compatibility between the images, if they use different versions of the standard library, or different bitness, for instance, the binaries will be incompatible.  In short -- build where you run.  One example: binaries built on Ubuntu 18.04 running on ARM architecture are incompatible with Ubuntu 18.04 running on a x86_64 architecture.
2. Yes, that's right.  One docker container for everything.
3. In that case, make sure to pass the absolute path to the shared library, not allow the OS to find it automatically, because that can cause all kinds of "creative" problems.  In your container, make sure that there are no unsatisfied dependencies on the librefprop.so. For instance: https://unix.stackexchange.com/questions/120015/how-to-find-out-the-dynamic-libraries-executables-loads-when-run
4. Make sure that ``var libpath = path.join(__dirname, '..', 'refprop', 'librefprop.so')`` gives a valid path to a file, and make sure to convert to an absolute path (use node to do this).  In Python it is ``os.path.abspath(path)``
5. Run the example C++ code that comes with REFPROP-headers in the same container where you built the shared library -- make sure that gives you the results you expect.  In the end you are not saving any time by trying to jump right to the end if the intermediate pieces don't work.  The key is to test that each piece is working, one at a time, to figure out where the problem is. With so many parts that have to fit together, its hard to know where to start to debug, which is why incremental development is so key.

**Comment by lee-zion** [comment]
1. ABI compatibility is the reason why I chose node-addon-api. nodejs provide ABI-stable API and that is N-API. node-addon-api is just a C++ Wrapper for N-API, so I think ABI compatibility might not be problem since I build ABI-stable code in the same OS.

2. Thanks. I'm just a very beginner of docker, so I should work on merging two dockerfiles now.

3, 4. Absolute path is already generated by node. `path.join` returns absolute path in node, as `os.path.abspath(path)` in Python.
```
var nodepath = path.join(__dirname, '..', 'docker-refprop.node')
// nodepath = /Users/username/REFPROP-debug/produce-node/basic-addon-api/docker-refprop.node
var libpath = path.join(__dirname, '..', 'refprop', 'librefprop.so')
// libpath = /Users/username/REFPROP-debug/produce-node/basic-addon-api/refprop/librefprop.so
```

5. Summarizing what you've suggested is
- Merge dockerfile for building shared library and node
- C++ code with shared library and REFPROP-Wrapper
- Above code with node

And I'll follow that way to debug. Thank you!

**Comment by ianhbell** [comment]
To my understanding, there are several layers of ABI compatibility, what you described is only ABI compatibility within node, but doesn't consider full-stack ABI compatibility.  This, among other reasons, is why it is sometimes very difficult to get binaries on linux to run on other machines.

**Comment by lee-zion** [comment]
I assumed that operating on same OS would prevent problems related to binary, but I now think that probably wrong.. complex binaries :(
I'll leave comment after make progress on this problem!

**Comment by ianhbell** [comment]
Good luck!  I'm happy to try to help as I can. The only "guaranteed" ABI solution is to do everything on the same machine.

**Comment by lee-zion** [comment]
Hi!
I'm so happy to come back here with a progress on the problem.

I've worked on merging two Dockerfiles, but it took too much time and still not finished.
I need to study more on Docker and Makefile to merge them.

So I've googled about my error: `libgfortran.so.4: cannot open shared object file: No such file or directory.`

And found [similar](https://stackoverflow.com/questions/50073888/libgfortran-so-4-cannot-open-shared-object-file-no-such-file-or-directory) [problems](https://forum.manjaro.org/t/r-error-while-loading-shared-libraries-libgfortran-so-4/57146/33) in python and R.

> ```
> RUN apt-get -y -m update && \
>     apt-get install -y cmake g++ python3 curl
> RUN curl -sL https://deb.nodesource.com/setup_10.x | bash
> RUN apt-get install -y nodejs
> ```
> Should I install something more to run shared library? (like gfortran or something?)
> 

When I install `gfortran` in Dockerfile of node(the same as REFPROP-cmake), everything went as I expected. I cannot ensure that I've perfectly solved the problem, but at least I can use it in my application now!
